compactcert/builder.go

```diff
@@ -0,0 +1,332 @@
+	}
+
+	hdrProto := config.Consensus[hdr.CurrentProtocol]
+	votersRnd := rnd.SubSaturate(basics.Round(hdrProto.CompactCertRounds))
+	votersHdr, err := ccw.ledger.BlockHdr(votersRnd)
+	if err != nil {
+		return builder{}, err
+	}
+
+	lookback := votersRnd.SubSaturate(basics.Round(hdrProto.CompactCertVotersLookback))
+	voters, err := ccw.ledger.CompactCertVoters(lookback)
+	if err != nil {
+		return builder{}, err
+	}
+
+	if voters == nil {
+		// Voters not tracked for that round.  Might not be a valid
+		// compact cert round; compact certs might not be enabled; etc.
+		return builder{}, fmt.Errorf("voters not tracked for lookback round %d", lookback)
+	}
+
+	p, err := ledger.CompactCertParams(votersHdr, hdr)
+	if err != nil {
+		return builder{}, err
+	}
+
+	var res builder
+	res.votersHdr = votersHdr
+	res.voters = voters
+	res.Builder, err = compactcert.MkBuilder(p, voters.Participants, voters.Tree)
+	if err != nil {
+		return builder{}, err
+	}
+
+	ccw.builders[rnd] = res
+	return res, nil
+}
+
+func (ccw *Worker) initBuilders() {
+	ccw.mu.Lock()
+	defer ccw.mu.Unlock()
+
+	var roundSigs map[basics.Round][]pendingSig
+	err := ccw.db.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+		roundSigs, err = getPendingSigs(tx)
+		return
+	})
+	if err != nil {
+		ccw.log.Warnf("initBuilders: getPendingSigs: %v", err)
+		return
+	}
+
+	for rnd, sigs := range roundSigs {
+		_, ok := ccw.builders[rnd]
+		if ok {
+			ccw.log.Warnf("initBuilders: round %d already present", rnd)
+			continue
+		}
+
+		builder, err := ccw.builderForRound(rnd)
+		if err != nil {
+			ccw.log.Warnf("initBuilders: builderForRound(%d): %v", rnd, err)
+			continue
+		}
+
+		for _, sig := range sigs {
+			pos, ok := builder.voters.AddrToPos[sig.signer]
+			if !ok {
+				ccw.log.Warnf("initBuilders: cannot find %v in round %d", sig.signer, rnd)
+				continue
+			}
+
+			err = builder.Add(pos, sig.sig, false)
+			if err != nil {
+				ccw.log.Warnf("initBuilders: cannot add %v in round %d: %v", sig.signer, rnd, err)
+				continue
+			}
+		}
+	}
+}
+
+func (ccw *Worker) handleSigMessage(msg network.IncomingMessage) network.OutgoingMessage {
+	var ssig sigFromAddr
+	err := protocol.Decode(msg.Data, &ssig)
+	if err != nil {
+		ccw.log.Warnf("ccw.handleSigMessage(): decode: %v", err)
+		return network.OutgoingMessage{Action: network.Disconnect}
+	}
+
+	fwd, err := ccw.handleSig(ssig, msg.Sender)
+	if err != nil {
+		ccw.log.Warnf("ccw.handleSigMessage(): %v", err)
+	}
+
+	return network.OutgoingMessage{Action: fwd}
+}
+
+func (ccw *Worker) handleSig(sfa sigFromAddr, sender network.Peer) (network.ForwardingPolicy, error) {
+	ccw.mu.Lock()
+	defer ccw.mu.Unlock()
+
+	builder, ok := ccw.builders[sfa.Round]
+	if !ok {
+		latest := ccw.ledger.Latest()
+		latestHdr, err := ccw.ledger.BlockHdr(latest)
+		if err != nil {
+			return network.Disconnect, err
+		}
+
+		if sfa.Round <= latestHdr.CompactCertLastRound {
+			// Already have a complete compact cert in ledger.
```

Comment:
Why are we sure we have this?

Reply:
`latestHdr.CompactCertLastRound` gets updated when that block contains a new compact cert transaction.  This means that blocks leading up to `latestHdr` must already store compact cert transactions for all blocks up to and including `latestHdr.CompactCertLastRound`.